### PR TITLE
add microseconds datetime format in views

### PR DESCRIPTION
**Pull request by Theodrem**

### CODE REVIEW

File: concrete_datastore/api/v1/views.py [hunk 1]
```diff
@@ -1842,7 +1843,7 @@ def check_date_format(date_type, param_values):
                         False,
                         Response(
                             data={
-                                'message': "Wrong date format, should be '{}' or '{}'".format(
+                                'message': "Wrong date format, should be '{}' or 'yyyy-mm-ddThh:mm:ss' or '{}'".format(
```
Comment (Souleymane-T) [inline concrete_datastore/api/v1/views.py:1846]: It should have been better to keep the format for

File: tests/tests_api_v1_0/test_filters.py [hunk 2]
```diff
@@ -334,9 +333,24 @@ def tearDown(self):
 
     def test_filter_range_date_and_datetime_fields(self):
         start_date = self.date.add(days=-1).to_date_string()
-        start_datetime = format_datetime(self.date.add(days=-1))
+        start_datetime = "2017-10-27T00:00:00Z"
+        print(start_datetime)
```
Comment (Souleymane-T) [inline tests/tests_api_v1_0/test_filters.py:337]: print to remove

File: tests/tests_api_v1_0/test_filters.py [hunk 3]
```diff
@@ -334,9 +333,24 @@ def tearDown(self):
 
     def test_filter_range_date_and_datetime_fields(self):
         start_date = self.date.add(days=-1).to_date_string()
-        start_datetime = format_datetime(self.date.add(days=-1))
+        start_datetime = "2017-10-27T00:00:00Z"
```
Comment (Souleymane-T) [inline tests/tests_api_v1_0/test_filters.py:336]: You can use the `self.date` variable set in the setUp here too

File: tests/tests_api_v1_0/test_filters.py [hunk 4]
```diff
@@ -334,9 +333,24 @@ def tearDown(self):
...
+        print(start_datetime)
         end_date = self.date.add(days=3).to_date_string()
-        end_datetime = format_datetime(self.date.add(days=3))
+        end_datetime = "2017-10-27T00:00:00Z"
```
Comment (Souleymane-T) [inline tests/tests_api_v1_0/test_filters.py:339]: You can use the `self.date` variable set in the setUp here too

File: tests/tests_api_v1_0/test_filters.py [hunk 5]
```diff
@@ -334,9 +333,24 @@ def tearDown(self):
...
+        )
+        resp = self.client.get(
+            url_date, HTTP_AUTHORIZATION="Token {}".format(self.token)
+        )
+        self.assertEqual(resp.status_code, status.HTTP_200_OK)
+
+    def test_filter_range_date_and_datetime_fields_with_microseconds(self):
+        start_date = self.date.add(days=-1).to_date_string()
+        start_datetime = "2017-10-27T00:00:00.000000Z"
+        print(start_datetime)
```
Comment (Souleymane-T) [inline tests/tests_api_v1_0/test_filters.py:351]: print to remove

File: tests/tests_api_v1_0/test_filters.py [hunk 6]
```diff
@@ -334,9 +333,24 @@ def tearDown(self):
 
     def test_filter_range_date_and_datetime_fields(self):
         start_date = self.date.add(days=-1).to_date_string()
-        start_datetime = format_datetime(self.date.add(days=-1))
+        start_datetime = "2017-10-27T00:00:00Z"
+        print(start_datetime)
         end_date = self.date.add(days=3).to_date_string()
-        end_datetime = format_datetime(self.date.add(days=3))
+        end_datetime = "2017-10-27T00:00:00Z"
+        url_date = '/api/v1/date-utc/?date__range={},{}&datetime__range={},{}'.format(
```
Comment (Souleymane-T) [inline tests/tests_api_v1_0/test_filters.py:340]: It's better to separate the tests and have one with the field `date` and another one with the field `datetime` only.  
You can still have a test with both after.

File: concrete_datastore/api/v1/views.py (hunk 1 above)
Comment (KhaledBousrih) [reply concrete_datastore/api/v1/views.py:1846]: or we can also define the `date_time_format` as 
```python
date_time_format = 'yyyy-mm-ddThh:mm:ss[.xxxxxx]Z'
```

(the `[` `]` to say that it's optional), and juste keen the same message as before, no ?

**Review by KhaledBousrih — CHANGES_REQUESTED** [review]
Can you also add a test case, where you filter a datetime with more than 6 milliseconds (that should raise a 400_BAD_REQUEST

File: concrete_datastore/api/v1/views.py (hunk 1 above)
Comment (Souleymane-T) [reply concrete_datastore/api/v1/views.py:1846]: Yes

**Comment by Souleymane-T** [comment]
Can you update the changelog please

File: tests/tests_api_v1_0/test_filters.py [hunk 7]
```diff
@@ -381,7 +912,7 @@ def setUp(self):
         self.confirmation = UserConfirmation.objects.create(user=self.user1)
         self.confirmation.confirmed = True
         self.confirmation.save()
-        url = '/api/v1/auth/login/'
+        url = '/api/v1.1/auth/login/'
```
Comment (KhaledBousrih) [inline tests/tests_api_v1_0/test_filters.py:915]: `api/v1/`

File: tests/tests_api_v1_0/test_filters.py [hunk 8]
```diff
@@ -423,7 +954,7 @@ def test_filter_objects_by_divider(self):
         self.assertEqual(results[0]['name'], "projet A")
 
         # User get only projects from cloisonY
-        url_filter = '/api/v1/project/?{}={}'.format(
+        url_filter = '/api/v1.1/project/?{}={}'.format(
```
Comment (KhaledBousrih) [inline tests/tests_api_v1_0/test_filters.py:957]: `api/v1/`

**Review by KhaledBousrih — CHANGES_REQUESTED** [review]

File: concrete_datastore/api/v1/filters.py [hunk 9]
```diff
@@ -29,6 +29,10 @@
 }
 
 JSON_FILTER_PATTERN = r'^\"(?P<str>.*)\"$|(?P<int>^\d+$)|(?P<float>^\d+\.\d+([e][+-]?\d+)?$)|(?P<bool>^true$|^false$)|(?P<null>^null$|^none$)'
+REGEX_DATETIME_MICROSECOND = (
+    "^\d{4}-\d{2}-\d{2}T\d{2}:\d{2}:\d{2}(\.\d{1,6})Z$"
```
Comment (KhaledBousrih) [inline concrete_datastore/api/v1/filters.py:33]: there is no need for a regex group here (`(\.\d{1,6})Z$"`)

you can remove the `(` `)`: `^\d{4}-\d{2}-\d{2}T\d{2}:\d{2}:\d{2}\.\d{1,6}Z$`

**Review by KhaledBousrih — CHANGES_REQUESTED** [review]

**Review by KhaledBousrih — APPROVED** [review]

**Pull request merged**